src/main/java/org/benetech/servicenet/adapter/icarol/RelationManager.java

```diff
@@ -73,30 +68,58 @@ private static void logProgress(int itemNr, int size) {
+        Set<Service> result = new HashSet<>();
         for (ICarolProgram program : relatedPrograms) {
-            persistence.importService(organization, importData, program).ifPresent(service ->
-                saveServiceRelatedData(location, program, service));
+            Optional<Service> service = getServiceToPersist(importData, program);
+            service.ifPresent(result::add);
         }
+        return result;
+    }
+
+    private Optional<Organization> importOrganization(ImportData importData, ICarolAgency agency,
+                                              Set<Location> locations, Set<Service> services) {
+        Organization extractedOrganization = mapper
+            .extractOrganization(agency, importData.getProviderName())
+            .sourceDocument(importData.getReport().getDocumentUpload());
+
+        extractedOrganization.setLocations(locations);
+        extractedOrganization.setServices(services);
+
+        return Optional.ofNullable(importManager
+            .createOrUpdateOrganization(extractedOrganization, agency.getId(), importData.getProviderName(),
+                importData.getReport()));
     }
 
-    private void saveServiceRelatedData(Location location, ICarolProgram program,
-                                        Service savedService) {
-        persistence.importEligibility(program, savedService);
+    private Optional<Location> getLocationToPersist(ImportData importData, ICarolSite site) {
+        return mapper.extractLocation(
+            site.getContactDetails(), site.getId(), importData.getProviderName())
+            .flatMap(extractedLocation -> {
+
+                extractedLocation.setPhysicalAddress(
+                    mapper.extractPhysicalAddress(site.getContactDetails()).orElse(null));
+
+                extractedLocation.setPostalAddress(
+                    mapper.extractPostalAddress(site.getContactDetails()).orElse(null));
 
-        persistence.importPhones(location, program, savedService);
-        persistence.importLangs(location, program, savedService);
-        persistence.importOpeningHours(program, savedService);
+                mapper.extractAccessibilityForDisabilities(site)
+                    .ifPresent(a -> extractedLocation.setAccessibilities(Set.of(a)));
+
+                return Optional.of(extractedLocation);
+            });
     }
 
-    private void saveEntitiesWithoutOrganization(ICarolDataToPersist data, ImportData importData) {
-        saveLocationsAndRelatedData(data.getSites(), data, importData, null, null);
-        saveServicesAndRelatedData(null, null, data.getPrograms(), importData);
+    private Optional<Service> getServiceToPersist(ImportData importData, ICarolProgram program) {
+        Service extractedService = mapper
+            .extractService(program, importData.getProviderName());
+
+        extractedService.setEligibility(mapper.extractEligibility(program).orElse(null));
```

Comment:
https://github.com/benetech/ServiceNet/wiki/iCarol-Data-Mapping-(Eden-and-UWBA)#restrictions

Eligibility cannot be null

Reply:
Indeed, that's why we set the whole Eligibility entity/relation to null. If mapper can't extract it, we shouldn't persist Empty eligibility.